services/device/geolocation/geoclue_location_provider.cc

```diff
@@ -0,0 +1,414 @@
+
+  // Invalidate weak pointers, so we don't continue any async operations.
+  weak_ptr_factory_.InvalidateWeakPtrs();
+
+  // Stop can be called before the gclue_client_ has resolved.
+  if (!gclue_client_) {
+    return;
+  }
+
+  dbus::MethodCall stop(kClientInterfaceName, "Stop");
+  gclue_client_->CallMethod(&stop, dbus::ObjectProxy::TIMEOUT_USE_DEFAULT,
+                            base::DoNothing());
+
+  // Reset pointers to dbus objects.
+  gclue_client_.reset();
+}
+
+const mojom::Geoposition& GeoClueLocationProvider::GetPosition() {
+  return last_position_;
+}
+
+void GeoClueLocationProvider::OnPermissionGranted() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  permission_granted_ = true;
+  StartClient();
+}
+
+void GeoClueLocationProvider::SetPosition(const mojom::Geoposition& position) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  last_position_ = position;
+
+  if (client_state_ != kStarted) {
+    return;
+  }
+
+  if (last_position_.error_code == mojom::Geoposition_ErrorCode::NONE &&
+      !device::ValidateGeoposition(last_position_)) {
+    return;
+  }
+  position_update_callback_.Run(this, last_position_);
+}
+
+void GeoClueLocationProvider::OnGetClientCompleted(dbus::Response* response) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  if (!response) {
+    SetPosition(GetErrorPosition());
+    return;
+  }
+
+  dbus::MessageReader reader(response);
+  dbus::ObjectPath path;
+  if (!reader.PopObjectPath(&path)) {
+    SetPosition(GetErrorPosition());
+    return;
+  }
+
+  gclue_client_ = bus_->GetObjectProxy(kServiceName, path);
+  SetClientProperties();
+}
+
+void GeoClueLocationProvider::SetClientProperties() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  auto* raw_ptr = new GeoClueProperties(gclue_client_.get());
+  std::unique_ptr<dbus::PropertySet> properties(raw_ptr);
+
+  // Wait for all properties to be set before starting the client.
+  const auto callback = base::BarrierCallback<bool>(
+      2, base::BindOnce(&GeoClueLocationProvider::OnSetClientProperties,
+                        weak_ptr_factory_.GetWeakPtr(), std::move(properties)));
+
+  raw_ptr->requested_accuracy_level.Set(
+      high_accuracy_requested_ ? kGeoClueAccuracyHigh : kGeoClueAccuracyLow,
+      callback);
+
+  raw_ptr->desktop_id.Set(kBraveDesktopId, callback);
+}
+void GeoClueLocationProvider::OnSetClientProperties(
+    std::unique_ptr<dbus::PropertySet> property_set,
+    std::vector<bool> success) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  if (base::ranges::any_of(success, [](const auto& value) { return !value; })) {
+    LOG(ERROR) << "Failed to set properties. GeoClue2 location provider will "
+                  "not work properly";
+    SetPosition(GetErrorPosition());
+    return;
+  }
+
+  ConnectSignal();
+}
+
+void GeoClueLocationProvider::ConnectSignal() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  gclue_client_->ConnectToSignal(
```

Comment:
need a null guard for `gclue_client_` here

Reply:
I don't think it can be |nullptr| here, unless `StopProvider` is called from a different thread (which should never happen)
1. `GetObjectProxy` never returns nullptr (https://source.chromium.org/chromium/chromium/src/+/main:dbus/bus.h;l=265)
2. `gclue_client_` is only reset in `StopProvider`, which calls `InvalidateWeakPtrs()` which should cancel the callback which calls this (it's only invoked via `OnSetClientProperties`, which is run via a callback with a `WeakPtr`).

I've added a `CHECK(gclue_client_)` to this method so we can be doubly sure, and make my assumption more explicit. My understanding of how things are interacting could be way off though :laughing: 

If you're wondering why `MaybeStartClient` does a null check, it's because that is also invoked via `OnPermissionGranted`, so the client may not have finished initializing. That's because we only want to start the client when it's finished initializing **AND** `OnPermissionGranted` has been called.

Reply:
if `ConnectSignal` is called after `StopProvider` (for instance in response to an async call to `OnSetClientProperties` then it would be null. Anytime you have an async call the the dbus thread, there could be another call to the provider on the main thread so the sequence on the main thread would be:
OnGetClientCompleted
SetClientProperties -> async call to dbus thread
StopProvider
return from dbus thread -> OnSetClientProperties
ConnectSignal -> Crash

Reply:
Okay, I've added the null check, but I'm still a little unsure how this happens - maybe my understanding of `WeakPtrs` is a bit broken?

This is what I thought would happen:
1. OnGetClientCompleted
2. SetClientProperties -> async call to dbus thread
3. StopProvider -> calls `InvalidateWeakPtrs()` and then resets `gclue_client_`
4. return from dbus thread -> The WeakPtr `OnSetClientProperties` is bound to has been invalidated, so it doesn't run.

But I guess that's wrong?

Reply:
I see, I missed that you were invalidating the weak ptr on this in StopProvider